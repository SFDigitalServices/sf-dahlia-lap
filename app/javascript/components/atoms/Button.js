import React from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import { toPx } from 'utils/cssUtils'

const buttonSizeStyles = (clearVerticalPadding, minWidth) => {
  const widthStyle = minWidth
    ? {
        minWidth: toPx(minWidth)
      }
    : {}

  const heightStyle = clearVerticalPadding
    ? {
        paddingTop: 0,
        paddingBottom: 0
      }
    : {}

  return {
    ...heightStyle,
    ...widthStyle
  }
}

const buttonChildrenStyles = () => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%'
})

const textStyles = (hasLeftIcon, hasRightIcon) => ({
  marginTop: '1rem',
  marginBottom: '1rem',
  marginRight: hasRightIcon && '1rem',
  marginLeft: hasLeftIcon && '1rem',
  flexGrow: 1
})

const iconWrapperStyles = () => ({
  flexShrink: 0,

  // we need to override font-size because of a css button selector that sets
  // the font size for all children. This is a problem because it adds 3px
  // to any icon wrapper.
  fontSize: 0
})

const wrapWithStyle = (node, style) => node && <span style={style}>{node}</span>

const Button = ({
  classes = null,
  iconLeft = null,
  iconRight = null,
  noBottomMargin,
  paddingHorizontal = 'normal',
  small = false,
  tertiary = false,
  text = null,
  tiny = false,
  type = 'button',
  minWidthPx = null,
  style = {},
  ...rest
}) => {
  const btnClassNames = classNames(classes, {
    'tight-padding': paddingHorizontal === 'tight',
    'extra-padding': paddingHorizontal === 'extra',
    small: small,
    tertiary: tertiary,
    tiny: tiny,
    'margin-bottom-none': noBottomMargin
  })

  const hasText = !!text

  return (
    <button
      className={btnClassNames}
      type={type}
      style={{ ...buttonSizeStyles(hasText, minWidthPx), ...style }}
      {...rest}
    >
      <div style={buttonChildrenStyles()}>
        {wrapWithStyle(iconLeft, iconWrapperStyles())}
        {wrapWithStyle(text, textStyles(!!iconLeft, !!iconRight))}
        {wrapWithStyle(iconRight, iconWrapperStyles())}
      </div>
    </button>
  )
}

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit']),
  noBottomMargin: PropTypes.bool,
  classes: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  iconLeft: PropTypes.node,
  iconRight: PropTypes.node,
  minWidthPx: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingHorizontal: PropTypes.oneOf(['normal', 'tight', 'extra']),
  small: PropTypes.bool,
  style: PropTypes.object,
  tertiary: PropTypes.bool,
  text: PropTypes.node,
  tiny: PropTypes.bool
}

export default Button
