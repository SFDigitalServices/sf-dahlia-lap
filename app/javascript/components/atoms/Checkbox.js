import React from 'react'

import classNames from 'classnames'
import { PropTypes } from 'prop-types'

const Checkbox = ({ id, checked = false, indeterminate = false, onClick = () => {} }) => (
  <>
    <input
      id={id}
      type='checkbox'
      checked={indeterminate || checked}
      onClick={onClick}
      // onChange handler required because we're setting 'checked' prop,
      // but we actually want our handler to be onClick instead,
      // so this is a no-op function.
      onChange={() => {}}
      className={classNames('no-margin', { indeterminate: indeterminate })}
    />
    <label style={{ margin: '0px' }} className='form-label' htmlFor={id} />
  </>
)

Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  indeterminate: PropTypes.bool,
  onClick: PropTypes.func
}
export default Checkbox
