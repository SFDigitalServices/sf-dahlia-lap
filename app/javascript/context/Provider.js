import React, { createContext, useReducer } from 'react'

import { createActions } from 'context/leaseUpActionCreators'
import Reducer, {
  getEmptyApplication,
  getEmptyApplicationsListData,
  getEmptyListing
} from 'context/Reducer'

const initialState = {
  breadcrumbData: {
    listing: getEmptyListing(),
    application: getEmptyApplication()
  },
  applicationsListData: getEmptyApplicationsListData()
}

export const AppContext = createContext()

const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(Reducer, initialState)
  const actions = createActions(dispatch)
  return <AppContext.Provider value={[state, actions]}>{children}</AppContext.Provider>
}

export default Provider
