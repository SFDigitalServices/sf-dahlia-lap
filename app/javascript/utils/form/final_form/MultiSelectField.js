import React from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'

import MultiSelect, { MultiSelectItemShape } from 'components/molecules/MultiSelect'

import { FieldError, HelpText, Label } from './Field'

const format = (options) => {
  return options && options.map((option) => (typeof option === 'string' ? option : option.value))
}

const MultiSelectField = ({
  blockNote,
  disabled = false,
  fieldName,
  helpText,
  id,
  label,
  options,
  onChange = () => {}
}) => (
  <Field name={fieldName} component={MultiSelect} format={format} formatOnBlur>
    {({ input, meta }) => {
      // React Select in some cases returns an empty string and MultiSelect accepts only arrays
      const value = input.value === '' ? null : input.value
      return (
        <>
          <div className={classNames('form-group', { error: meta.error && meta.touched })}>
            <Label
              label={label}
              id={id || `form-${fieldName}`}
              fieldName={fieldName}
              blockNote={blockNote}
            />
            <MultiSelect
              {...input}
              options={options}
              value={value}
              onChange={(event) => {
                input.onChange(event)
                onChange && onChange(event)
              }}
              disabled={disabled}
            />
            <FieldError meta={meta} />
            {helpText && <HelpText note={helpText} describeId={`describe-${id || fieldName}`} />}
          </div>
        </>
      )
    }}
  </Field>
)

MultiSelectField.propTypes = {
  blockNote: PropTypes.string,
  disabled: PropTypes.bool,
  fieldName: PropTypes.string.isRequired,
  helpText: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(MultiSelectItemShape).isRequired
}

export default MultiSelectField
