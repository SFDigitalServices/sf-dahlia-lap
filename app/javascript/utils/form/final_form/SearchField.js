import React, { useRef } from 'react'
import PropTypes from 'prop-types'

import { Field } from 'react-final-form'
import classNames from 'classnames'
import { Label, FieldError, HelpText } from './Field'

const SearchField = ({
  blockNote,
  disabled = false,
  fieldName,
  format,
  formatOnBlur = false,
  helpText,
  id,
  isDirty = false,
  label,
  maxLength,
  onClearClick,
  placeholder,
  validation
}) => {
  const inputElement = useRef(null)

  return (
    <Field
      name={fieldName}
      validate={validation}
      format={format}
      formatOnBlur={formatOnBlur && isDirty}
    >
      {({ input, meta }) => {
        const hasValue = !!input.value
        return (
          <>
            <div
              className={classNames('search', {
                'form-group': label,
                error: meta.error && meta.touched
              })}
            >
              <Label
                label={label}
                id={id || `form-${fieldName}`}
                fieldName={fieldName}
                blockNote={blockNote}
              />
              <input
                {...input}
                id={id || `form-${fieldName}`}
                className={classNames({ 'has-value': hasValue, error: meta.error && meta.touched })}
                type='text'
                maxLength={maxLength}
                placeholder={placeholder}
                disabled={disabled}
                ref={inputElement}
              />
              {hasValue && (
                <button
                  className='ui-medium i-aluminum button-link search-icon'
                  onClick={() => {
                    onClearClick()
                    inputElement.current.focus()
                  }}
                  type='button'
                >
                  <svg>
                    <use xlinkHref={`#i-close-round`} />
                  </svg>
                </button>
              )}
              <FieldError meta={meta} />
              {helpText && <HelpText note={helpText} describeId={`describe-${id || fieldName}`} />}
            </div>
          </>
        )
      }}
    </Field>
  )
}

SearchField.propTypes = {
  blockNote: PropTypes.string,
  disabled: PropTypes.bool,
  fieldName: PropTypes.string.isRequired,
  format: PropTypes.func,
  formatOnBlur: PropTypes.bool,
  helpText: PropTypes.string,
  id: PropTypes.string,
  isDirty: PropTypes.bool,
  label: PropTypes.string,
  maxLength: PropTypes.number,
  onClearClick: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  validation: PropTypes.func
}
export default SearchField
