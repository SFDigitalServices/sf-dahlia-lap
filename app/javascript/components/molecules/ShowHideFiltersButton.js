import React from 'react'

import PropTypes from 'prop-types'

import Button from '../atoms/Button'
import StyledIcon from '../atoms/StyledIcon'

const getNumFiltersAppliedIcon = (numFiltersApplied) => {
  if (numFiltersApplied < 0 || numFiltersApplied > 9) {
    console.error(`Cannot display a number > 9 or < 0 in filter bubble: ${numFiltersApplied}`)
    numFiltersApplied = 9
  }

  return <StyledIcon icon={`filter-qty--${numFiltersApplied}`} />
}

const ShowHideFiltersButton = ({
  isShowingFilters = false,
  numFiltersApplied = 0,
  onClick = () => {}
}) => {
  const showFiltersCount = numFiltersApplied > 0
  return (
    <Button
      iconLeft={<StyledIcon icon='list-unordered' customSizeRem='1.25rem' />}
      iconRight={showFiltersCount ? getNumFiltersAppliedIcon(numFiltersApplied) : null}
      text={isShowingFilters ? 'Hide Filters' : 'Show Filters'}
      tightPadding
      // Set a min width so that there is no jitter when toggling between
      // show and hide text
      minWidthPx={showFiltersCount ? '236px' : '195px'}
      onClick={onClick}
    />
  )
}

ShowHideFiltersButton.propTypes = {
  isShowingFilters: PropTypes.bool,
  numFiltersApplied: PropTypes.number,
  onClick: PropTypes.func
}

export default ShowHideFiltersButton
