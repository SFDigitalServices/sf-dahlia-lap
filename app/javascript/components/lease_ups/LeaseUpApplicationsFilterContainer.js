import React, { useState } from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Form } from 'react-final-form'

import Button from 'components/atoms/Button'
import Checkbox from 'components/atoms/Checkbox'
import { LEASE_UP_APPLICATION_FILTERS } from 'components/lease_ups/applicationFiltersConsts'
import LeaseUpApplicationsFilters from 'components/lease_ups/LeaseUpApplicationsFilters'
import Loading from 'components/molecules/Loading'
import ShowHideFiltersButton from 'components/molecules/ShowHideFiltersButton'
import StatusDropdown from 'components/molecules/StatusDropdown'
import { useAppContext } from 'utils/customHooks'
import SearchField from 'utils/form/final_form/SearchField'
import formUtils from 'utils/formUtils'

const styles = {
  marginBottomZero: {
    marginBottom: 0
  },
  bulkEditCheckbox: {
    display: 'flex',
    alignItems: 'center',
    marginBottom: '1.25rem',
    paddingRight: '1rem'
  }
}

const getNumFiltersApplied = (form) => {
  return LEASE_UP_APPLICATION_FILTERS.filter((f) => {
    const value = form.getState().values[f.fieldName]
    return !(!value || value.length === 0)
  }).length
}

const LeaseUpApplicationsFilterContainer = ({
  onSubmit,
  preferences = [],
  loading = false,
  bulkCheckboxesState = [],
  onBulkLeaseUpStatusChange,
  onBulkLeaseUpCommentChange,
  onClearSelectedApplications = () => {},
  onSelectAllApplications = () => {}
}) => {
  const [isShowingFilters, setIsShowingFilters] = useState(false)
  const [hasChangedFilters, setHasChangedFilters] = useState(false)

  const [
    {
      applicationsListData: { appliedFilters }
    }
  ] = useAppContext()

  const onClickShowHideFilters = () => setIsShowingFilters(!isShowingFilters)

  const handleFilterChange = () => {
    setHasChangedFilters(true)
  }

  const handleFormSubmit = (filters, form) => {
    onSubmit(formUtils.scrubEmptyValues(filters, true))
    form.resetFieldState('search')
    setHasChangedFilters(false)
  }

  const handleClearFilters = (form) => {
    LEASE_UP_APPLICATION_FILTERS.forEach((f) => {
      const isChanged = !!form.getState().values[f.fieldName]
      if (isChanged) {
        form.change(f.fieldName, null)
        setHasChangedFilters(true)
      }
    })
  }

  const applicationIds = Object.entries(bulkCheckboxesState)
  const numChecked = applicationIds.filter(([_, checked]) => checked).length
  const allChecked = applicationIds.length > 0 && applicationIds.length === numChecked

  const handleCheckboxClicked = () => {
    if (numChecked > 0) {
      onClearSelectedApplications()
    } else {
      onSelectAllApplications()
    }
  }

  return (
    <Loading isLoading={loading}>
      <Form
        onSubmit={handleFormSubmit}
        initialValues={appliedFilters}
        render={({ form, handleSubmit }) => (
          <form style={styles.marginBottomZero} onSubmit={handleSubmit} noValidate>
            <div className='filter-row'>
              <div className='filter-group filter-group--left'>
                <div style={styles.bulkEditCheckbox}>
                  <Checkbox
                    id='bulk-edit-controller'
                    indeterminate={!allChecked && numChecked > 0}
                    checked={allChecked}
                    onClick={handleCheckboxClicked}
                  />
                </div>
                <div className='filter-group_action'>
                  <div className='padding-right--half d-inline-block'>
                    <StatusDropdown
                      classes={{ tertiary: numChecked === 0 }}
                      disabled={numChecked === 0}
                      onChange={onBulkLeaseUpStatusChange}
                      minWidthPx={'185px'}
                      placeholder={'Set Status'}
                      forceDisplayPlaceholderText
                      tertiary={numChecked === 0}
                    />
                  </div>
                  <Button
                    tertiary={numChecked === 0}
                    disabled={numChecked === 0}
                    onClick={onBulkLeaseUpCommentChange}
                    type='button'
                    minWidthPx='185px'
                    text='Add Comment'
                    textAlign='left'
                    noBottomMargin
                  />
                </div>
              </div>
              <div className='filter-group'>
                <div className='filter-group_item__large'>
                  <SearchField
                    onClearClick={() => form.change('search', '')}
                    fieldName='search'
                    id='test-search'
                    placeholder='Application, First Name, Last Name...'
                    minWidth='300px'
                  />
                </div>
                <div className='filter-group_action'>
                  <Button
                    className={classNames('primary', {
                      'tertiary-inverse': !form.getFieldState('search')?.modified
                    })}
                    text='Search'
                    type='submit'
                  />
                </div>
                <div className='filter-group_action'>
                  <ShowHideFiltersButton
                    isShowingFilters={isShowingFilters}
                    onClick={onClickShowHideFilters}
                    numFiltersApplied={getNumFiltersApplied(form)}
                  />
                </div>
              </div>
            </div>
            {isShowingFilters && (
              <LeaseUpApplicationsFilters
                preferences={preferences}
                hasChangedFilters={hasChangedFilters}
                onFilterChange={handleFilterChange}
                onClearFilters={() => handleClearFilters(form)}
              />
            )}
          </form>
        )}
      />
    </Loading>
  )
}

LeaseUpApplicationsFilterContainer.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onBulkLeaseUpStatusChange: PropTypes.func.isRequired,
  onBulkLeaseUpCommentChange: PropTypes.func.isRequired,
  onClearSelectedApplications: PropTypes.func,
  onSelectAllApplications: PropTypes.func,
  preferences: PropTypes.array,
  loading: PropTypes.bool,
  bulkActionApplications: PropTypes.object
}

export default LeaseUpApplicationsFilterContainer
