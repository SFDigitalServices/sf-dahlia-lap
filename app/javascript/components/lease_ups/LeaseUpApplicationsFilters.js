import React from 'react'

import classNames from 'classnames'

import Button from 'components/atoms/Button'
import { COLORS } from 'components/atoms/colors'
import { LEASE_UP_APPLICATION_FILTERS } from 'components/lease_ups/applicationFiltersConsts'
import FormGrid from 'components/molecules/FormGrid'
import MultiSelectField from 'utils/form/final_form/MultiSelectField'

const styles = {
  containerEndJustified: {
    display: 'flex',
    justifyContent: 'flex-end'
  },
  whiteSmokeBackground: {
    background: COLORS.snow
  }
}

const LeaseUpApplicationsFilters = ({
  preferences = [],
  hasChangedFilters,
  onFilterChange = () => {},
  onClearFilters = () => {}
}) => {
  const renderFilter = (filter) => (
    <FormGrid.Item width='25%' key={filter.fieldName}>
      <MultiSelectField
        options={filter.getOptions({ listingPreferences: preferences })}
        fieldName={filter.fieldName}
        label={filter.label}
        placeholder={filter.placeholder}
        onChange={onFilterChange}
      />
    </FormGrid.Item>
  )

  return (
    <div className='padding-top--2x padding-left padding-right' style={styles.whiteSmokeBackground}>
      <FormGrid.Row expand paddingBottom>
        {LEASE_UP_APPLICATION_FILTERS.map((f) => renderFilter(f))}
      </FormGrid.Row>

      <div className='small-12 margin-top' style={styles.containerEndJustified}>
        <Button
          type='submit'
          classes={classNames('margin-right--half', {
            'tertiary-inverse': !hasChangedFilters,
            primary: hasChangedFilters
          })}
          paddingHorizontal='extra'
          text='Apply filters'
        />
        <Button
          classes='tertiary'
          paddingHorizontal='extra'
          text='Clear all'
          onClick={onClearFilters}
        />
      </div>
    </div>
  )
}
export default LeaseUpApplicationsFilters
