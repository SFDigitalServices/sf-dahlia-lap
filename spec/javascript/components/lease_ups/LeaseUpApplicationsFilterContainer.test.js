import React from 'react'

import { mount } from 'enzyme'
import { act } from 'react-dom/test-utils'

import Button from 'components/atoms/Button'
import Checkbox from 'components/atoms/Checkbox'
import LeaseUpApplicationsFilterContainer from 'components/lease_ups/LeaseUpApplicationsFilterContainer'
import LeaseUpApplicationsFilters from 'components/lease_ups/LeaseUpApplicationsFilters'
import ShowHideFiltersButton from 'components/molecules/ShowHideFiltersButton'
import Provider from 'context/Provider'
import MultiSelectField from 'utils/form/final_form/MultiSelectField'
import SearchField from 'utils/form/final_form/SearchField'

const mockSubmit = jest.fn()
const mockOnClearSelectedApplications = jest.fn()
const mockOnSelectAllApplications = jest.fn()

const getNode = (bulkCheckboxesState = {}) => (
  <Provider>
    <LeaseUpApplicationsFilterContainer
      onSubmit={mockSubmit}
      onClearSelectedApplications={mockOnClearSelectedApplications}
      onSelectAllApplications={mockOnSelectAllApplications}
      bulkCheckboxesState={bulkCheckboxesState}
      onBulkLeaseUpStatusChange={() => {}}
      onBulkLeaseUpCommentChange={() => {}}
      preferences={['pref option 1', 'pref option 2']}
    />
  </Provider>
)

const getWrapper = (bulkCheckboxesState = {}) => mount(getNode(bulkCheckboxesState))

describe('LeaseUpApplicationsFilterContainer', () => {
  describe('status checkbox', () => {
    let wrapper
    describe('when the list of applications is empty', () => {
      beforeEach(() => {
        wrapper = getWrapper({})
      })

      test('should render the checkbox as unchecked', () => {
        expect(wrapper.find(Checkbox).props().checked).toBeFalsy()
      })

      test('should render the checkbox as not indeterminate', () => {
        expect(wrapper.find(Checkbox).props().indeterminate).toBeFalsy()
      })

      test('should call select all on click', () => {
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
        wrapper.find(Checkbox).simulate('click')
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(1)
      })
    })

    describe('when the list of applications is all unchecked', () => {
      beforeEach(() => {
        wrapper = getWrapper({ appId1: false, appId2: false })
      })

      test('should render the checkbox as unchecked', () => {
        expect(wrapper.find(Checkbox).props().checked).toBeFalsy()
      })

      test('should render the checkbox as not indeterminate', () => {
        expect(wrapper.find(Checkbox).props().indeterminate).toBeFalsy()
      })

      test('should call select all on click', () => {
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
        wrapper.find(Checkbox).simulate('click')
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(1)
      })
    })
    describe('when the list of applications has some checked and some unchecked', () => {
      beforeEach(() => {
        wrapper = getWrapper({ appId1: false, appId2: true })
      })

      test('should render the checkbox as checked=false', () => {
        expect(wrapper.find(Checkbox).props().checked).toBeFalsy()
      })

      test('should render the checkbox with indeterminate=true', () => {
        expect(wrapper.find(Checkbox).props().indeterminate).toBeTruthy()
      })

      test('should call clear all on click', () => {
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
        wrapper.find(Checkbox).simulate('click')
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(1)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
      })
    })
    describe('when the list of applications is all checked', () => {
      beforeEach(() => {
        wrapper = getWrapper({ appId1: true, appId2: true })
      })

      test('should render the checkbox as checked=false', () => {
        expect(wrapper.find(Checkbox).props().checked).toBeTruthy()
      })

      test('should render the checkbox with indeterminate=false', () => {
        expect(wrapper.find(Checkbox).props().indeterminate).toBeFalsy()
      })

      test('should call clear all on click', () => {
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(0)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
        wrapper.find(Checkbox).simulate('click')
        expect(mockOnClearSelectedApplications.mock.calls).toHaveLength(1)
        expect(mockOnSelectAllApplications.mock.calls).toHaveLength(0)
      })
    })
  })

  describe('search field', () => {
    test('it does not show the clear button when there is no text in the field', () => {
      const wrapper = getWrapper()
      expect(wrapper.exists('.search-icon')).toBe(false)
    })
    test('it renders a clear input icon when there is text in the search input', () => {
      const wrapper = getWrapper()
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(wrapper.exists('.search-icon')).toBe(true)
    })
    test('the clear input button clears the input', () => {
      const wrapper = getWrapper()
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(wrapper.find(SearchField).find('input').props().value).toEqual('my search query')

      wrapper.find('.search-icon').simulate('click')
      expect(wrapper.find(SearchField).find('input').props().value).toEqual('')
    })
    test('the search field is tertiary-inverse unless the field has been modified', () => {
      const wrapper = getWrapper()
      const graySearchButtonWrapper = (wrapper) =>
        wrapper.find(Button).find({ text: 'Search' }).find('button.tertiary-inverse')

      // search button is disabled on first render
      expect(graySearchButtonWrapper(wrapper)).toHaveLength(1)

      // search button should be enabled after we modify the field
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my search query' } })
      wrapper.update()
      expect(graySearchButtonWrapper(wrapper)).toHaveLength(0)

      // search button should be disabled after we submit
      wrapper.find(Button).find({ text: 'Search' }).find('button').simulate('submit')
      expect(graySearchButtonWrapper(wrapper)).toHaveLength(1)

      // search button should be re-enabled if we modify after submit
      wrapper
        .find('.search')
        .find('input')
        .simulate('change', { target: { value: 'my other search query' } })
      wrapper.update()
      expect(graySearchButtonWrapper(wrapper)).toHaveLength(0)
    })

    describe('filters button and panel', () => {
      let wrapper
      beforeEach(() => {
        wrapper = getWrapper()
      })

      test('should default to hiding the filters', () => {
        expect(wrapper.find(ShowHideFiltersButton).props().isShowingFilters).toBeFalsy()
        expect(wrapper.find(LeaseUpApplicationsFilters)).toHaveLength(0)
      })

      describe('when the show/hide filters button is clicked', () => {
        beforeEach(async () => {
          await act(async () => wrapper.find(ShowHideFiltersButton).simulate('click'))
          wrapper.update()
        })

        test('should show the filters panel', () => {
          expect(wrapper.find(ShowHideFiltersButton).props().isShowingFilters).toBeTruthy()
          expect(wrapper.find(LeaseUpApplicationsFilters)).toHaveLength(1)
        })

        test('should should set hasChangedFilters = false', () => {
          expect(wrapper.find(LeaseUpApplicationsFilters).props().hasChangedFilters).toBeFalsy()
        })

        describe('when filters are cleared', () => {
          beforeEach(async () => {
            await act(async () => wrapper.find(LeaseUpApplicationsFilters).props().onClearFilters())
            wrapper.update()
          })

          test('should not should set hasChangedFilters = true if filters were already cleared', () => {
            expect(wrapper.find(LeaseUpApplicationsFilters).props().hasChangedFilters).toBeFalsy()
          })
        })

        describe('when a filter is changed', () => {
          beforeEach(async () => {
            await act(async () => wrapper.find(MultiSelectField).at(0).invoke('onChange')())
            wrapper.update()
          })

          test('should should set hasChangedFilters = true', () => {
            expect(wrapper.find(LeaseUpApplicationsFilters).props().hasChangedFilters).toBeTruthy()
          })

          describe('when filters are cleared', () => {
            beforeEach(async () => {
              await act(async () =>
                wrapper.find(LeaseUpApplicationsFilters).props().onClearFilters()
              )
              wrapper.update()
            })

            test('should keep hasChangedFilters = true', () => {
              expect(
                wrapper.find(LeaseUpApplicationsFilters).props().hasChangedFilters
              ).toBeTruthy()
            })
          })
        })
      })
    })
  })
})
