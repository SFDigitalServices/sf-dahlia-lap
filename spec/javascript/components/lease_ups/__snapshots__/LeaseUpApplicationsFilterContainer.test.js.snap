// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaseUpApplicationsFilterContainer it matches the expected snapshot with default preferences 1`] = `
<ContextProvider
  value={
    Object {
      "batch": [Function],
      "blur": [Function],
      "change": [Function],
      "destroyOnUnregister": false,
      "focus": [Function],
      "getFieldState": [Function],
      "getRegisteredFields": [Function],
      "getState": [Function],
      "initialize": [Function],
      "isValidationPaused": [Function],
      "mutators": Object {},
      "pauseValidation": [Function],
      "registerField": [Function],
      "reset": [Function],
      "resetFieldState": [Function],
      "restart": [Function],
      "resumeValidation": [Function],
      "setConfig": [Function],
      "submit": [Function],
      "subscribe": [Function],
    }
  }
>
  <form
    noValidate={true}
    onSubmit={[Function]}
    style={
      Object {
        "marginBottom": 0,
      }
    }
  >
    <div
      className="filter-row"
    >
      <div
        className="filter-group filter-group--left"
      >
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "marginBottom": "1.25rem",
              "paddingRight": "1rem",
            }
          }
        >
          <Checkbox
            checked={false}
            id="bulk-edit-controller"
            indeterminate={false}
            onClick={[Function]}
          />
        </div>
        <div
          className="filter-group_action"
        >
          <StatusDropdown
            disabled={true}
            forceDisplayPlaceholderText={true}
            minWidthPx="185px"
            onChange={[Function]}
            placeholder="Set Status"
          />
        </div>
      </div>
      <div
        className="filter-group"
      >
        <div
          className="filter-group_item__large"
        >
          <SearchField
            fieldName="search"
            id="test-search"
            minWidth="300px"
            onClearClick={[Function]}
            placeholder="Application, First Name, Last Name..."
          />
        </div>
        <div
          className="filter-group_action"
        >
          <Button
            className="primary tertiary-inverse"
            text="Search"
            type="submit"
          />
        </div>
        <div
          className="filter-group_action"
        >
          <ShowHideFiltersButton
            isShowingFilters={false}
            numFiltersApplied={0}
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </form>
</ContextProvider>
`;
